Added checks for existing content is git repo

oatmealraisin opened the pull request:
This fixes an error when the git repo is empty, e.g. the user only ran
`git init`.  If the repo is empty, we use system copy commands instead
of git commands, and warn the user.

Also added warnings for if there is a .git directory in the project, but
a Git binary is not installed.

Currently working on tests, just posting this for discussion
@bparees 

Old output, git installed:
```

---> Installing application source ...
ERROR: cp: cannot stat '/tmp/src/.': No such file or directory

ERROR: An error occurred: non-zero (13) exit code from openshift/ruby-22-centos7
```

Old output, git not installed:
```

---> Installing application source ...
---> Building your Ruby application from source ...
WARNING: Rubygem Rack is not installed in the present image.
Add rack to your Gemfile in order to start the web server.

```

New output, git installed:
```
Warning: file:///home/rymurphy/documents/projects/go/src/github.com/openshift/source-to-image/asdf is a Git repository, but is empty.

---> Installing application source ...
---> Building your Ruby application from source ...
WARNING: Rubygem Rack is not installed in the present image.
Add rack to your Gemfile in order to start the web server.

```

New output, git not installed
```
Warning: file:///home/rymurphy/documents/projects/go/src/github.com/openshift/source-to-image/asdf is a Git repository, but you do not have git installed.
Warning: file:///home/rymurphy/documents/projects/go/src/github.com/openshift/source-to-image/asdf is a Git repository, but is empty.

---> Installing application source ...
---> Building your Ruby application from source ...
WARNING: Rubygem Rack is not installed in the present image.
Add rack to your Gemfile in order to start the web server.

```
Fixes #298

oatmealraisin commented:
Made the "Warning: " all caps

bparees commented on pkg/scm/git/git.go:
```diff
@@ -196,8 +196,19 @@ func ParseURL(source string) (*url.URL, error) {
+
+	if isLocalGitRepository(source) {
+		if !hasGitBinary() {
+			fmt.Println("WARNING: " + source + " is a Git repository, but you do not have git installed.")
```
glog.Warning ?

bparees commented on pkg/scm/git/git.go:
```diff
@@ -196,8 +196,19 @@ func ParseURL(source string) (*url.URL, error) {
+		}
+
+		if isEmptyGitRepository(source) {
+			fmt.Println("WARNING: " + source + " is a Git repository, but is empty.")
```
glog.Warning?

bparees replied:
tell them to use --copy if they want to use the local files instead.

bparees replied:
or rather, tell them you *are* using --copy, since that is what is happening.

bparees commented on pkg/scm/git/git.go:
```diff
@@ -379,6 +390,27 @@ func doesExist(dir string) bool {
+	gitDir := strings.TrimPrefix(dir, "file://") + "/.git/"
+
+	// Search the content of the .git directory for content
+	cmd := "find " +
```
this isn't going to work on windows, needs a more portable solution.

php-coder commented on pkg/scm/git/git.go:
```diff
@@ -196,8 +196,19 @@ func ParseURL(source string) (*url.URL, error) {
+		}
+	}
+
+	return !isLocalGitRepository(source) || !hasGitBinary() || isEmptyGitRepository(source)
```
I see that we're executing `isLocalGitRepository()` twice always and also executing `hasGitBinary()` and `isEmptyGitRepository()` twice sometimes. Because they're doing IO, maybe it's better to execute them only once? I mean, let's extract them to variables.

oatmealraisin replied:
I agree, I'll change that

oatmealraisin commented:
[test]
@bparees ptal?

php-coder commented on pkg/scm/git/git.go:
```diff
@@ -379,6 +395,33 @@ func doesExist(dir string) bool {
+
+	gitDir := strings.TrimPrefix(dir, "file://") + "/.git"
+	// Search the content of the .git directory for content
+	directories := [2]string{gitDir + "/objects", gitDir + "/refs"}
```
Could using of `/` lead to a problem on Windows?

php-coder replied:
BTW, instead of prepending `gitDir` to each element of the array here, we can prepend it later, when we're passing first argument to `filepath.Walk()`. It's up to you, of course, it's just idea.

bparees replied:
yeah this should be using filepath.Join

oatmealraisin replied:
noted, thanks

php-coder commented on pkg/scm/git/git.go:
```diff
@@ -379,6 +395,33 @@ func doesExist(dir string) bool {
+					// If we find a file, the git directory is "not empty"
+					// We're looking for object blobs, and ref files
+					if !info.IsDir() {
+						isEmpty = false
```
If we found a file do we need to continue or one file is enough? In this case we can put `break` here, right?

php-coder replied:
BTW, one more idea is to don't ignore `err` argument. I'm not sure how to use it in this case, probably, we could write it and at the end if `isEmpty` is `true` then show a warning with this error.

oatmealraisin replied:
Correct, we only need 1 file to confirm it's a git repo.

As for the error, I don't know if we should do error checking in such a simple function.  Maybe we could combine the isLocalGitRepo() and isEmptyGitRepo() to isValidGitRepo(), with an extra error return value?

bparees replied:
yeah you can just bubble the error up and ultimately fail the whole thing.

oatmealraisin replied:
I'll combine them and do some error handling, this would keep us from assuming --copy as well :)

php-coder commented on pkg/scm/scm_test.go:
```diff
@@ -14,7 +14,12 @@ func createLocalGitDirectory(t *testing.T) string {
+	os.MkdirAll(filepath.Join(dir, ".git/branches"), 0777)
+	os.MkdirAll(filepath.Join(dir, ".git/objects/foo"), 0777)
+	os.Create(filepath.Join(dir, ".git/objects/foo") + "bar")
+	//os.Create(filepath.Join(dir, ".git/refs/heads") + "master")
```
I hope that we need this commented code for something..

oatmealraisin replied:
Thanks for the feedback!  I forgot to take this out.  Since we already create a `bar` object, it is a "valid" git repository, because that would represent a commit.  I'll add a comment explaining that

php-coder commented:
I've reviewed a bit, but the last opinion on @bparees of course, so you can ignore my comments, but I would be happy if some of ideas will help somehow.

bparees commented on pkg/scm/git/git.go:
```diff
@@ -196,8 +196,24 @@ func ParseURL(source string) (*url.URL, error) {
-	return !isLocalGitRepository(source) || !hasGitBinary()
+	isEmpty := isEmptyGitRepository(source)
+	isLocal := isLocalGitRepository(source)
+	hasGit := hasGitBinary()
```
you can optimize this further i think by lazily initialized the variables as needed inside the isLocal if block

eg if it's not a local git repo, you don't care if "hasGit" or "isEmpty" are true.

oatmealraisin replied:
Would it be OK if I got rid of that glog, or split it up?  I'll put something up

bparees replied:
yeah you can rework it.

oatmealraisin commented:
Updated to bubble errors up and stop the program.  

New output: Git installed
```
s2i build file://$PWD openshift/ruby-22-centos7 newapp
ERROR: An error occurred: The git repository "file:///home/rymurphy/documents/projects/go/src/github.com/openshift/source-to-image/asdf" has no tracking information or commits
```
New output: Git not installed
```
 s2i build file://$PWD openshift/ruby-22-centos7 newapp
ERROR: An error occurred: Git was not found in the PATH
```
@bparees  @php-coder ptal?

php-coder commented on pkg/scm/git/clone.go:
```diff
@@ -23,7 +23,9 @@ func (c *Clone) Download(config *api.Config) (*api.SourceInfo, error) {
-	if c.ValidCloneSpec(config.Source) {
+	if ok, err := c.ValidCloneSpec(config.Source); err != nil {
+		return nil, err
+	} else if ok {
```
Small neat: we don't need `else` if we're returning from the first branch.

oatmealraisin replied:
Good point, will change

php-coder commented on pkg/scm/git/git.go:
```diff
@@ -217,83 +219,95 @@ func makePathAbsolute(source string) string {
-		if !doesExist(path) {
+	if doesExist(path) {
+		var isValidGit bool
+		isValidGit, err = isValidGitRepository(path)
```
You can write it in a one line, right?
```go
isValidGit, err := isValidGitRepository(path)
```

oatmealraisin replied:
This causes an error with go because it thinks we're shadowing the named return value err :(

php-coder replied:
Ah, ok.

php-coder commented on pkg/scm/git/git.go:
```diff
@@ -492,3 +556,27 @@ func pipeToLog(reader io.Reader, log func(...interface{})) {
+
+func (err InvalidReferenceError) Error() string {
+	return fmt.Sprintf("Invalid reference \"%s\" in %s", err.ref, err.source)
+}
```
I'm not sure, but maybe it will be better to move them to `pkg/errors/errors.go`?

oatmealraisin replied:
I'm ok with either!  Is that where all errors go?  It would make sense either way

php-coder replied:
As I said, I'm not sure how it should be. Will wait @bparees response.

bparees replied:
yeah it would make sense to move them there.

oatmealraisin replied:
Will do

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,83 +219,95 @@ func makePathAbsolute(source string) string {
 // binary, and if a frag/ref has a bad format
-func ParseFile(source string) (details *FileProtoDetails, mods *URLMods) {
+func ParseFile(source string) (details *FileProtoDetails, mods *URLMods, err error) {
+	// NOTE: Why don't we just trim the source here?
```
not sure what you mean. we're explicitly needing to track the difference between something like:
file://bparees@github.com  (user specified a file with an unusual name)
and
bparees@github.com (user specified an ssh syntax w/ no explicit protocol)

what are you proposing w/ trim?

oatmealraisin replied:
You understood correctly, I was wondering what the use-cases for file:// were.

Within ParseFile, whenever we use `path` we trim the prefix before using it.  If we trimmed it at the beginning, we wouldn't have to make those checks further down.  

Since the function name states we're working with a file, would it make sense to just trim the file:// prefix outright?  We'd still be tracking its use with `protoSpecified`

bparees replied:
six of one, half dozen of the other.

oatmealraisin replied:
I'm not sure what you mean :(

bparees replied:
"these two things are basically the same so it doesn't really matter"

you're welcome to change it, but if you update fileExists to not strip file://, you need to make sure no one who is calling it expects it to retain that behavior.

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,83 +219,95 @@ func makePathAbsolute(source string) string {
+
+		if refSpecified {
+			// if ref/frag bad, return bad
+			// NOTE: Should we throw an error here?
```
no, we're already returning BadRef in the struct, assume the caller is responsible for handling that if it cares.

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,83 +219,95 @@ func makePathAbsolute(source string) string {
-		// if ref/frag bad, return bad
-		if !gitSSHURLPathRef.MatchString(ref) {
+
+		if refSpecified {
```
i don't see the purpose of this new if statement.  if the ref is "" it will always be valid anyway, so the existing logic should work, no?

oatmealraisin replied:
It's more for the good return value.  We could put the error check outside of the if statement, but we'd have to keep the good return inside of it.

bparees replied:
what error check?

oatmealraisin replied:
On line 262, ```if !gitSSHURLPathRef.MatchString(ref)```

oatmealraisin replied:
I see what you're talking about now, sorry for the confusion

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,83 +219,95 @@ func makePathAbsolute(source string) string {
+			}
+
+			// return good
+			isValidGit, _ := isValidGitRepository(path)
```
why are you checking this again when you already did it on 247?

bparees commented on pkg/scm/scm.go:
```diff
@@ -23,9 +23,13 @@ func DownloaderForSource(s string, forceCopy bool) (build.Downloader, string, er
+		return nil, s, err
+	}
 
+	// NOTE: This could be handled by an error
```
i think it's being handled correctly, it's up to the caller whether it wants to try to press on w/ the other git information returned and ignore the bad ref, or bail out.

oatmealraisin commented:
After adding the errors to our errors package, the new outputs follow the format

```
s2i build file:///$PWD openshift/ruby-22-centos7 newapp
ERROR: An error occurred: The git repository "file:////home/rymurphy/documents/projects/go/src/github.com/openshift/source-to-image/asdf" has no tracking information or commits
ERROR: Suggested solution: Either remove the .git directory from the project, or update its tracking information.
ERROR: If the problem persists consult the docs at https://github.com/openshift/source-to-image/tree/master/docs. Eventually reach us on freenode #openshift or file an issue at https://github.com/openshift/source-to-image/issues providing us with a log from your build using --loglevel=3
```
@bparees What do you think about this output?

bparees commented:
> ERROR: Suggested solution: Either remove the .git directory from the project, or update its tracking information.


the suggested solutions would be any of:

1) commit the local files to the repository
2) remove the .git folder
3) use --copy instead

oatmealraisin commented:
New output
```
ERROR: An error occurred: The git repository "file:///home/rymurphy/documents/projects/go/src/github.com/openshift/source-to-image/asdf" has no tracking information or commits
ERROR: Suggested solution: Either commit files to the Git repository, remove the .git directory from the project, or use --copy to ignore the repository.
ERROR: If the problem persists consult the docs at https://github.com/openshift/source-to-image/tree/master/docs. Eventually reach us on freenode #openshift or file an issue at https://github.com/openshift/source-to-image/issues providing us with a log from your build using --loglevel=3
```

Also had to make changes to make sure --copy would work even with a faulty Git
@bparees ptal?

bparees commented on pkg/errors/errors.go:
```diff
@@ -232,3 +234,24 @@ func NewUserNotAllowedError(image string, onbuild bool) error {
+	return Error{
+		Message:    "Git was not found in the PATH",
+		ErrorCode:  MissingGitBinaryError,
+		Suggestion: "Install Git from your distribution's vendor",
```
or use --copy to ignore the git repository

oatmealraisin replied:
@bparees  Good catch, updated

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,65 +220,57 @@ func makePathAbsolute(source string) string {
+
 		// if ref/frag bad, return bad
-		if !gitSSHURLPathRef.MatchString(ref) {
+		if refSpecified && !gitSSHURLPathRef.MatchString(ref) {
```
you're only going to get here if err!=nil above, is that intentional?  doesn't seem right.

oatmealraisin replied:
Yes I believe so.  Before we were working with just the path, but now we're checking to see if the user included a valid reference to a branch.  

We could just add `!gitSSHURLPathRef...` in the struct declaration below, 
```
 		details = &FileProtoDetails{
 			FileExists:     true,
			UseCopy:        !isValidGit,
 			ProtoSpecified: protoSpecified,
 			BadRef:         !gitSSHURLPathRef.MatchString(ref),
 		}
```
and let the caller figure out that the source input has a bad ref.

oatmealraisin commented:
@bparees I've update isValidGit to not check for a Git binary, and moved that check up to ParseFile.  I've also removed the predeclared returns, and updated test cases. ptal?

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,84 +220,75 @@ func makePathAbsolute(source string) string {
+	// switch to
+	if doesExist(path) {
+		isValidGit, err := isValidGitRepository(path)
+		hasGit := hasGitBinary()
```
no reason to check this if isValidGit is false, right?

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,84 +220,75 @@ func makePathAbsolute(source string) string {
+	if doesExist(path) {
+		isValidGit, err := isValidGitRepository(path)
+		hasGit := hasGitBinary()
+		if err != nil {
```
shouldn't this be:  if err!=nil || !isValidGit || !hasGit {

?

under any of those circumstances we should be returning UseCopy: true (and returning the err, if any, so the caller can assess if it wants to proceed)

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,84 +220,75 @@ func makePathAbsolute(source string) string {
 				FileExists:     true,
 				ProtoSpecified: protoSpecified,
-				BadRef:         true,
+				UseCopy:        !(isValidGit && hasGit),
```
and then this can be UseCopy: false  since you've eliminated the "invalid git repo" and "missing git binary" cases above, if you make the change i suggested.

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,84 +220,75 @@ func makePathAbsolute(source string) string {
 			}
-			mods = nil
-			return
+			return details, nil, nil
```
I think these two blocks where you're setting up details for the case where we do have a valid git repo and git binary should be combined:
```
badRef:=refSpecified && !gitSSHURLPathRef.MatchString(ref)
details:= &FileProtoDetails {
				UseCopy:        false,
				FileExists:     false,
				ProtoSpecified: protoSpecified,
				BadRef:         badRef,
}
mods= &URLMods{
			Scheme: "file",
			Path:   makePathAbsolute(strings.TrimPrefix(source, "file://")),
			Ref:    ref,
		}
return details, mods, nil
```
The old code returned mods as "nil" in the case that there was a BadRef, but it's not obvious to me that it matters whether mods gets returned as nil, or gets returned with the proper value.  I think the assumption the code made was that if BadRef was true, the caller wasn't going to do anything further.  But if that's true, it's fine to actually return a valid mods also.

oatmealraisin commented:
@bparees I've made the ParseFile function more lazy in it's checks, ptal?

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,84 +220,98 @@ func makePathAbsolute(source string) string {
 	// in each valid case, like the prior logic in scm.go did, we'll make the
 	// paths absolute and prepend file:// to the path which callers should
 	// switch to
+	if doesExist(path) {
```
this is what i was trying to guide you towards.   I think it retains all the logic we've discussed, but is more compact:

```
	// in each valid case, like the prior logic in scm.go did, we'll make the
	// paths absolute and prepend file:// to the path which callers should
	// switch to
	if doesExist(path) {

		// Is there even a valid .git repository?
		isValidGit, err := isValidGitRepository(path)
		hasGit:=false
		if isValidGit {
		  hasGit=hasGitBinary()
		}
		// err indicates empty git repo (should also consider throwing an error for no .git folder
		// so the caller knows specifically why it's not a valid git repo)
		if err != nil || !isValidGit || !hasGit {
			details := &FileProtoDetails{
				UseCopy:        true,
				FileExists:     true,
				BadRef:         false,  
				ProtoSpecified: protoSpecified,
			}
			mods := &URLMods{
				Scheme: "file",
				Path:   makePathAbsolute(strings.TrimPrefix(path, "file://")),
				Ref:    ref,
			}
			return details, mods, err
		}

		// Check is the #ref is valid
		badRef:=refSpecified && !gitSSHURLPathRef.MatchString(ref)

		details := &FileProtoDetails{
			BadRef:         badRef,
			FileExists:     true,
			ProtoSpecified: protoSpecified,
			// this value doesn't really matter, we should not proceed if the git ref is bad
			// but let's fallback to "copy" mode if the ref is invalid.
			UseCopy:        badRef,  
		}

		mods := &URLMods{
			Scheme: "file",
			Path:   makePathAbsolute(strings.TrimPrefix(path, "file://")),
			Ref:    ref,
		}
		return details, mods, nil
	}

	// File does not exist, return bad
	details := &FileProtoDetails{
		UseCopy:        false,
		FileExists:     false,
		BadRef:         false,
		ProtoSpecified: protoSpecified,
	}
	return details, nil, nil
```

oatmealraisin commented:
@bparees I've put in your suggestion, ptal?

bparees commented on pkg/scm/git/git.go:
```diff
@@ -217,84 +220,81 @@ func makePathAbsolute(source string) string {
-			details = &FileProtoDetails{
-				UseCopy:        false,
+		// err indicates empty git repo (should also consider throwing an error for no .git folder
+		// so the caller knows specifically why it's not a valid git repo)
```
this comment was a suggestion to you.  either implement it or remove the comment, but it doesn't belong here (if you're not going to do it now, but still think it's a good idea, it should be to a //TODO on isValidGitRepository

bparees commented:
> @bparees I've put in your suggestion, ptal?

did it make sense?

oatmealraisin commented:
@bparees I decided not to implement an error for missing .git

bparees commented on pkg/scm/git/git.go:
```diff
@@ -366,11 +365,52 @@ func ParseSSH(source string) (*URLMods, error) {
-func isLocalGitRepository(dir string) bool {
-	_, err := os.Stat(fmt.Sprintf("%s/.git", strings.TrimPrefix(dir, "file://")))
-	return !(err != nil && os.IsNotExist(err))
+// isValidgitRepository checks to see if there is a .git repository in the
```
isValidGitRepository

oatmealraisin replied:
Good catch

bparees commented on pkg/scm/git/git_test.go:
```diff
@@ -13,23 +13,56 @@ import (
-	os.RemoveAll(filepath.Join(d, ".git"))
-	if isLocalGitRepository(d) == true {
+
+	os.RemoveAll(d)
```
why is this one not a defer before you do the test?

oatmealraisin replied:
A bit of confusion about how defer works, will change

bparees commented on pkg/scm/git/git_test.go:
```diff
@@ -13,23 +13,56 @@ import (
+	// There are no tracking objects in the .git repository
+	// Should return true with an error
+	if ok, err := isValidGitRepository(d); ok == false || err == nil {
+		t.Errorf("The %q directory is not a git repository", d)
```
for completeness you should check that the error you got is the type of error (emptygitdir) you expected.

bparees commented on pkg/scm/git/git_test.go:
```diff
@@ -13,23 +13,56 @@ import (
+
+	// There is no .git in the provided directory
+	// Should return false with no error
+	if ok, _ := isValidGitRepository(d); ok == true {
```
you should check that err==nil

bparees commented on pkg/scm/scm.go:
```diff
@@ -23,8 +23,11 @@ func DownloaderForSource(s string, forceCopy bool) (build.Downloader, string, er
-	details, mods := git.ParseFile(s)
+	details, mods, err := git.ParseFile(s)
 	glog.V(4).Infof("return from ParseFile file exists %v proto specified %v use copy %v", details.FileExists, details.ProtoSpecified, details.UseCopy)
+	if err != nil && !forceCopy {
```
this should be checking the type of error that came back.  the only error you can ignore here is an "EmptyGitDir" error, right?

oatmealraisin commented:
@bparees I added some checks for error types within the tests, as well as to scm.go, ptal?

bparees commented on pkg/scm/git/git_test.go:
```diff
@@ -10,28 +10,80 @@ import (
+	}
+
+	d = createEmptyLocalGitDirectory(t)
+	defer os.RemoveAll(d)
```
actually now i'm wondering how this will work also.  is it going to call RemoveAll with the current value of "d" at the time of the return, or is it going to call RemoveAll with the value of "d" at the time the defer was declared?  Obviously we need the latter behavior, or you need to use separate variables for each.

oatmealraisin replied:
That's what I was wondering before :) I looked it up, defer evaluates the arguments when called but delays the actual function until the surrounding function returns.

bparees commented:
one last question on the behavior of defer, and needs commit squashing.

oatmealraisin commented:
I'll squash now

bparees commented:
[merge]

php-coder commented on pkg/errors/errors.go:
```diff
@@ -232,3 +234,24 @@ func NewUserNotAllowedError(image string, onbuild bool) error {
+
+// NewMissingGitBinaryError returns a new error that indicates that we couldn't
+// find a Git binary in the path.
+func NewMissingGitBinaryError() error {
```
@oatmealraisin Ryan, this method looks unused. Should we use it somewhere or just remove?

oatmealraisin replied:
@php-coder We could use it when hasGitBinary() is callled on line 251 of git/git.go, but I think we talked about whether we should error out when the user doesn't have Git installed and decided that we would just use --copy instead.  

Are we tight about removing unused code?  Imo it's a legitimate error, and we may use it if we decide otherwise.  I'm not very opposed to just nixing it, thought.

bparees replied:
since we don't have immediate plans to resume using this, we should remove it.
